README.md
# Invalidator for AWS S3-based Cloudfront distributions

[![Build Status](https://secure.travis-ci.org/laurilehmijoki/cf-s3-invalidator.png)]
(http://travis-ci.org/laurilehmijoki/cf-s3-invalidator)

If your Amazon Web Services Cloudfront distribution is based on AWS S3, then
this library may be useful to you.

This library fetches all the object keys of a particular S3 bucket and then
calls on them the invalidation REST API of Cloudfront.

## Install

`gem install cf-s3-invalidator`

## Usage

You can specify the configuration as CLI parameters:

`cf-s3-inv --key <AWS-KEY> --secret <AWS-SECRET> --distribution <CLOUDFRONT-DISTRIBUTION-ID>`

Or you can store them into the file *_cf_s3_invalidator.yml*:

    aws_key: YOUR_AWS_ACCESS_KEY_ID
    aws_secret: YOUR_AWS_SECRET_ACCESS_KEY
    cloudfront_distribution_id: YOUR_CLOUDFRONT_DISTRIBUTION_ID

Then you can just run:

`cf-s3-inv`

## Development

Run tests:

`rake test`

## Troubleshooting

* Verify that the AWS user has CloudFront permissions. In the AWS web interface, this
  will be in the `permissions` section of 
  `https://console.aws.amazon.com/iam/home?region=REGION-CODE#users/USERNAME`

* If you're using cf-s3-invalidator from Travis CI, be sure the credentials are
  [encrypted](http://docs.travis-ci.com/user/encryption-keys/) with the correct key
  for the repository

## License

MIT


changelog.md
# Changelog

## 0.4.2

* Stop using Ruby 1.8.7; Nokogiri needs >=1.9.2
* Change deprecated OpenSSL::Digest::Digest to OpenSSL::Digest 

## 0.4.1

* Print a nicer-looking invalidation report
* Start using [Semantic Versioning](http://semver.org/)


lib/cf-s3-invalidator.rb
require 'rubygems'
require 'openssl'
require 'digest/sha1'
require 'net/https'
require 'base64'
require 'colored'

module CloudfrontS3Invalidator
  class CloudfrontClient
    def initialize(aws_account, aws_secret, distribution)
      @aws_account = aws_account
      @aws_secret = aws_secret
      @distribution = distribution
    end

    def get_s3_bucket_name
      res = sign_and_call(
        "https://cloudfront.amazonaws.com/2012-05-05/distribution/#{@distribution}",
        Net::HTTP::Get)
      matches =
        res.body.scan(/<DomainName>([\w|\.]+)\.s3\.amazonaws\.com<\/DomainName>/)
      if matches.empty?
        nil
      else
        s3_bucket_name = matches.first.first
        s3_bucket_name
      end
    end

    private

    def sign_and_call(url, method, body = nil)
      date = Time.now.strftime("%a, %d %b %Y %H:%M:%S %Z")
      digest = Base64.encode64(
        OpenSSL::HMAC.digest(OpenSSL::Digest.new('sha1'), @aws_secret, date)).strip
      uri = URI.parse(url)
      req = method.new(uri.path)
      req.initialize_http_header({
        'x-amz-date' => date,
        'Content-Type' => 'text/xml',
        'Authorization' => "AWS %s:%s" % [@aws_account, digest]
      })
      req.body = body unless body == nil
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = true
      http.verify_mode = OpenSSL::SSL::VERIFY_NONE
      res = http.request(req)
      if res.code.to_i.between? 200, 299
        res
      else
        raise "AWS API call failed. Reason:".red + "\n" + res.body
      end
    end
  end
end


lib/cloudfront_invalidator.rb
require 'rubygems'
require 'openssl'
require 'digest/sha1'
require 'net/https'
require 'base64'
require 'colored'

# Adapted from:
# Confabulus @ http://blog.confabulus.com/2011/05/13/cloudfront-invalidation-from-ruby
class CloudfrontInvalidator
  def initialize(aws_account, aws_secret, distribution)
    @aws_account = aws_account
    @aws_secret = aws_secret
    @distribution = distribution
  end

  def invalidate(items)
    items = to_cf_keys(items)
    date = Time.now.strftime("%a, %d %b %Y %H:%M:%S %Z")
    digest = Base64.encode64(
      OpenSSL::HMAC.digest(OpenSSL::Digest::Digest.new('sha1'), @aws_secret, date)).strip
    uri = URI.parse(
      "https://cloudfront.amazonaws.com/2012-05-05/distribution/#{@distribution}/invalidation")
    req = Net::HTTP::Post.new(uri.path)
    req.initialize_http_header({
      'x-amz-date' => date,
      'Content-Type' => 'text/xml',
      'Authorization' => "AWS %s:%s" % [@aws_account, digest]
    })
    req.body = %|
      <InvalidationBatch>
        <Paths>
          <Quantity>#{items.length}</Quantity>
          <Items>
            #{to_xml items}
          </Items>
        </Paths>
        <CallerReference>#{Time.now.utc.to_i}</CallerReference>
      </InvalidationBatch>
    |
    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = true
    http.verify_mode = OpenSSL::SSL::VERIFY_NONE
    res = http.request(req)
    print_operation_result(res, items)
  end

  def print_operation_result(http_response, items)
    success = http_response.code == '201'
    puts "Invalidating Cloudfront items".cyan
    items.each do |item|
      puts "  #{item}".blue
    end
    if success
      puts "succeeded".green
    else
      puts "FAILED, reason:".red
      puts http_response.body
    end
  end

  def to_cf_keys(s3_keys)
    s3_keys.map { |s3_key| "/#{s3_key}" }
  end

  def to_xml(items)
    items.map { |item| "<Path>#{item}</Path>" }
  end
end


lib/cloudfront_client.rb
require 'rubygems'
require 'openssl'
require 'digest/sha1'
require 'net/https'
require 'base64'
require 'colored'

class CloudfrontClient
  def initialize(aws_account, aws_secret, distribution)
    @aws_account = aws_account
    @aws_secret = aws_secret
    @distribution = distribution
  end

  def invalidate(items)
    items = to_cf_keys(items)
    body = %|
      <InvalidationBatch>
        <Paths>
          <Quantity>#{items.length}</Quantity>
          <Items>
            #{to_xml items}
          </Items>
        </Paths>
        <CallerReference>#{Time.now.utc.to_i}</CallerReference>
      </InvalidationBatch>
    |
    res = sign_and_call(
      "https://cloudfront.amazonaws.com/2012-05-05/distribution/#{@distribution}/invalidation",
      Net::HTTP::Post,
      body)
    print_invalidation_result(res, items)
  end

  def get_s3_bucket_name
    res = sign_and_call(
      "https://cloudfront.amazonaws.com/2012-05-05/distribution/#{@distribution}",
      Net::HTTP::Get)
    matches =
      res.body.scan(/<DomainName>([\w|\.]+)\.s3\.amazonaws\.com<\/DomainName>/)
    if matches.empty?
      nil
    else
      s3_bucket_name = matches.first.first
      s3_bucket_name
    end
  end

  private

  def sign_and_call(url, method, body = nil)
    date = Time.now.strftime("%a, %d %b %Y %H:%M:%S %Z")
    digest = Base64.encode64(
      OpenSSL::HMAC.digest(OpenSSL::Digest::Digest.new('sha1'), @aws_secret, date)).strip
    uri = URI.parse(url)
    req = method.new(uri.path)
    req.initialize_http_header({
      'x-amz-date' => date,
      'Content-Type' => 'text/xml',
      'Authorization' => "AWS %s:%s" % [@aws_account, digest]
    })
    req.body = body unless body == nil
    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = true
    http.verify_mode = OpenSSL::SSL::VERIFY_NONE
    res = http.request(req)
    if res.code.to_i.between? 200, 299
      res
    else
      raise "AWS API call failed. Reason:".red + "\n" + res.body
    end
  end

  def print_invalidation_result(http_response, items)
    puts "Invalidating Cloudfront items..."
    items.each do |item|
      puts "  #{item}".yellow
    end
    puts "succeeded".green
  end

  def to_cf_keys(s3_keys)
    s3_keys.map { |s3_key| "/#{s3_key}" }
  end

  def to_xml(items)
    items.map { |item| "<Path>#{item}</Path>" }
  end
end
